src/Shared/NamedPipeUtil.cs

```diff
@@ -25,5 +35,251 @@ internal static string GetPipeNameOrPath(string pipeName)
                 return pipeName;
             }
         }
+#if !FEATURE_PIPEOPTIONS_CURRENTUSERONLY
+        //  This code needs to be in a separate method so that we don't try (and fail) to load the Windows-only APIs when JIT-ing the code
+        //  on non-Windows operating systems
+        private static void ValidateRemotePipeSecurityOnWindows(NamedPipeClientStream nodeStream)
+        {
+            SecurityIdentifier identifier = WindowsIdentity.GetCurrent().Owner;
+#if FEATURE_PIPE_SECURITY
+            PipeSecurity remoteSecurity = nodeStream.GetAccessControl();
+#else
+            var remoteSecurity = new PipeSecurity(nodeStream.SafePipeHandle, System.Security.AccessControl.AccessControlSections.Access |
+                System.Security.AccessControl.AccessControlSections.Owner | System.Security.AccessControl.AccessControlSections.Group);
+#endif
+            IdentityReference remoteOwner = remoteSecurity.GetOwner(typeof(SecurityIdentifier));
+            if (remoteOwner != identifier)
+            {
+                CommunicationsUtilities.Trace("The remote pipe owner {0} does not match {1}", remoteOwner.Value, identifier.Value);
+                throw new UnauthorizedAccessException();
+            }
+        }
+#endif
+
+        /// <summary>
+        /// Attempts to connect to the specified process.
+        /// </summary>
+        internal static NamedPipeClientStream TryConnectToProcess(int nodeProcessId, int timeout, Handshake? handshake)
+        {
+            string pipeName = GetPipeNameOrPath("MSBuild" + nodeProcessId);
+            return TryConnectToProcess(pipeName, nodeProcessId, timeout, handshake);
+        }
+
+        /// <summary>
+        /// Attempts to connect to the specified process.
+        /// </summary>
+        internal static NamedPipeClientStream TryConnectToProcess(string pipeName, int timeout, Handshake? handshake)
+        {
+            return TryConnectToProcess(pipeName, null, timeout, handshake);
+        }
+
+        private static NamedPipeClientStream TryConnectToProcess(string pipeName, int? nodeProcessId, int timeout, Handshake? handshake)
+        {
+            NamedPipeClientStream nodeStream = new NamedPipeClientStream(".", pipeName, PipeDirection.InOut, PipeOptions.Asynchronous
+#if FEATURE_PIPEOPTIONS_CURRENTUSERONLY
+                                                                         | PipeOptions.CurrentUserOnly
+#endif
+                                                                         );
+            if (nodeProcessId.HasValue)
+            {
+                CommunicationsUtilities.Trace("Attempting connect to PID {0} with pipe {1} with timeout {2} ms", nodeProcessId.Value, pipeName, timeout);
+            }
+            else
+            {
+                CommunicationsUtilities.Trace("Attempting connect to process {0} with pipe {1} with timeout {2} ms", pipeName, pipeName, timeout);
+            }
```

Comment:
(I know you just moved this, but:)
```suggestion
           CommunicationsUtilities.Trace("Attempting connect to PID {0} with pipe {1} with timeout {2} ms", nodeProcessId.HasValue ? nodeProcessId.Value : pipeName, pipeName, timeout);
```